library/core/src/cmp.rs

```diff
@@ -58,10 +58,18 @@ use self::Ordering::*;
 ///
 /// - **Transitive**: if `A: PartialEq<B>` and `B: PartialEq<C>` and `A:
 ///   PartialEq<C>`, then **`a == b` and `b == c` implies `a == c`**.
+///   This must also work for longer chains, such as when `A: PartialEq<B>`, `B: PartialEq<C>`,
+///   `C: PartialEq<D>`, and `A: PartialEq<D>` all exist.
 ///
 /// Note that the `B: PartialEq<A>` (symmetric) and `A: PartialEq<C>`
 /// (transitive) impls are not forced to exist, but these requirements apply
 /// whenever they do exist.
+/// The requirement for transitive chains has a subtle consequence: if a crate `a` with type `A`
+/// importing types `B1`, `B2` from another crate `b` defines instances of both `A: PartialEq<B1>`
+/// and `B2: PartialEq<A>`, then if `b` adds a `B1: PartialEq<B2>` implementation in the future that
+/// could create new violations of transitivity. Effectively, crate `a` is making `B1` and `B2`
+/// comparable even though `b` does not permit comparing them. Crate `a` should generally refrain
+/// from doing that to avoid the potential for such conflicts.
```

Comment:
Now on third-or-so read, I do get the feeling that that one sentence is unnecessarily nested/complex. It could more plainly (and a bit more explicitly) be something like

> […] has a subtle consequence: Consider a crate `alpha` defining type `A`, and a crate `beta` defining types `B1` and `B2`. If `alpha` uses `beta` as a dependency and provides trait implementations for `B1: PartialEq<A>` and `A: PartialEq<B2>`, the transitivity rules don’t apply, as an implementation `B1: PartialEq<B2>` does not exist; but at a later point `beta` could decide to add such a `B1: PartialEq<B2>` implementation, in a way that would then make `alpha`’s traits implementations violate transitivity. Effectively, `alpha` is already making `B1` and `B2` comparable for equality even though `beta` does not permit comparing them. Crate `alpha` should […]

Also, writing this I noticed that your `A: PartialEq<B1>` and `B2: PartialEq<A>` was the wrong way around, unable to be violating `B1: PartialEq<B2>`.

<details><summary>screenshot, and also markdown source for easy copying</summary>

![Screenshot_20230831_013330](https://github.com/rust-lang/rust/assets/3986214/3bb4cdbe-5d20-432a-9cb9-2292dc6e6174)

```md
[…] has a subtle consequence: Consider a crate `alpha` defining type `A`, and a crate `beta` defining types `B1` and `B2`. If `alpha` uses `beta` as a dependency and provides trait implementations for `B1: PartialEq<A>` and `A: PartialEq<B2>`, the transitivity rules don't apply, as an implementation `B1: PartialEq<B2>` does not exist; but at a later point `beta` could decide to add such a `B1: PartialEq<B2>` implementation, in a way that would then make `alpha`'s traits implementations violate transitivity. Effectively, `alpha` is already making `B1` and `B2` comparable for equality even though `beta` does not permit comparing them. Crate `alpha` should […]
```

</details>

Reply:
Thanks! I have adopted and tweaked your wording. I also swapped the crate names since I found it more natural that alpha would come before beta in the dependency chain.